### PR TITLE
[BATCH-2675] Explicitly considering Start Time not being null for running Job Executions

### PR DESCRIPTION
**Pull request by dimitrisli**
In an attempt to exclude erroneous Job Executions whenever considering Running Job Executions (i.e. via `JobExplorer#findRunningJobExecutions()`). 

An example is whenever a `TaskRejectedException` is thrown after submitting to the `taskExecutor` in `SimpleJobLauncher#run()`, the `JobExecution` is left without a _Start_ or _End Time_ and therefore being considered as a Running Job Execution while it's not. 

Also related tests are fixed.

### CODE REVIEW

File: spring-batch-core/src/main/java/org/springframework/batch/core/JobExecution.java [hunk 1]
```diff
@@ -234,10 +234,10 @@ public StepExecution createStepExecution(String stepName) {
 	 * be noted that this does not necessarily mean that it has been persisted
 	 * as such yet.
 	 *
-	 * @return true if the end time is null
+	 * @return true if the end time is null while the start time is not null
```
Comment (benas) [inline spring-batch-core/src/main/java/org/springframework/batch/core/JobExecution.java]: I would replace "while" with "and" here.

File: spring-batch-core/src/test/java/org/springframework/batch/core/JobExecutionTests.java [hunk 2]
```diff
@@ -34,8 +35,13 @@
  */
 public class JobExecutionTests {
 
-	private JobExecution execution = new JobExecution(new JobInstance(new Long(11), "foo"),
-			new Long(12), new JobParameters(), null);
+	private JobExecution execution;
+
+	@Before
```
Comment (benas) [inline spring-batch-core/src/test/java/org/springframework/batch/core/JobExecutionTests.java]: we can keep the inline initialization in this PR. I usually don't change (or polish) and add code in the same change set (unless it is necessary) in order to see what really fixes the issue. Moving the initialization to a `@Before` method (which is a good idea) is better done in a polishing commit.

File: spring-batch-core/src/test/java/org/springframework/batch/core/JobExecutionTests.java [hunk 3]
```diff
@@ -250,4 +258,9 @@ public void testFailureExceptions() {
 		assertTrue(allExceptions.contains(exception));
 		assertTrue(allExceptions.contains(stepException1));
 	}
+
+	private void simulatingARunningExecution() {
```
Comment (benas) [inline spring-batch-core/src/test/java/org/springframework/batch/core/JobExecutionTests.java]: We can inline this method then remove it. This:

```java
@Test
public void testIsRunning() {
   execution.setStartTime(new Date());
   assertTrue(execution.isRunning());
   execution.setEndTime(new Date(100L));
   assertFalse(execution.isRunning());
}
```

is IMO more coherent than this:

```java
@Test
public void testIsRunning() {
   simulatingARunningExecution();
   assertTrue(execution.isRunning());
   execution.setEndTime(new Date(100L));
   assertFalse(execution.isRunning());
}
```

File: spring-batch-core/src/test/java/org/springframework/batch/core/repository/dao/AbstractJobExecutionDaoTests.java [hunk 4]
```diff
@@ -194,14 +194,23 @@ public void testGetMissingLastExecution() {
 	@Transactional
 	@Test
 	public void testFindRunningExecutions() {
-
+		//Normally completed JobExecution as EndTime is populated
```
Comment (benas) [inline spring-batch-core/src/test/java/org/springframework/batch/core/repository/dao/AbstractJobExecutionDaoTests.java]: I would add a start time here as it makes no sense to have an execution with an end time but without a start time.  For example:

```java
JobExecution exec = new JobExecution(jobInstance, jobParameters);
exec.setCreateTime(new Date(0));
exec.setStartTime(new Date(1L));
exec.setEndTime(new Date(2L));
```

File: spring-batch-core/src/test/java/org/springframework/batch/core/repository/dao/AbstractJobExecutionDaoTests.java [hunk 5]
```diff
@@ -194,14 +194,23 @@ public void testGetMissingLastExecution() {
...
 		JobExecution exec = new JobExecution(jobInstance, jobParameters);
 		exec.setCreateTime(new Date(0));
 		exec.setEndTime(new Date(1L));
 		exec.setLastUpdated(new Date(5L));
 		dao.saveJobExecution(exec);
 
+		//BATCH-2675
+		//Abnormal JobExecution as both StartTime and EndTime are null
```
Comment (benas) [inline spring-batch-core/src/test/java/org/springframework/batch/core/repository/dao/AbstractJobExecutionDaoTests.java]: Thanks for adding this! It illustrates the issue we are addressing 👍

File: spring-batch-core/src/test/java/org/springframework/batch/core/explore/support/MapJobExplorerFactoryBeanTests.java [hunk 6]
```diff
@@ -36,8 +39,12 @@
 	public void testCreateExplorer() throws Exception {
 
 		MapJobRepositoryFactoryBean repositoryFactory = new MapJobRepositoryFactoryBean();
-		repositoryFactory.getObject().createJobExecution("foo", new JobParameters());
-		
+		JobExecution jobExecution = repositoryFactory.getObject().createJobExecution("foo", new JobParameters());
+
+		//simulating a running job execution
+		jobExecution.setStartTime(new Date());
+		repositoryFactory.getObject().update(jobExecution);
```
Comment (benas) [inline spring-batch-core/src/test/java/org/springframework/batch/core/explore/support/MapJobExplorerFactoryBeanTests.java]: The two calls to `repositoryFactory.getObject();` in this method can be extracted to a variable.

File: spring-batch-core/src/test/java/org/springframework/batch/core/jsr/launch/JsrJobOperatorTests.java [hunk 7]
```diff
@@ -677,4 +677,10 @@ public JsrJobOperator jobOperator(JobExplorer jobExplorer, JobRepository jobrepo
 			return new JsrJobOperator(jobExplorer, jobrepository, jobParametersConverter, transactionManager);
 		}
 	}
+
+	private JobExecution simulatingARunningJobExecution() {
```
Comment (benas) [inline spring-batch-core/src/test/java/org/springframework/batch/core/jsr/launch/JsrJobOperatorTests.java]: This method can be inlined and removed.

**Comment by benas** [comment]
Hi @dimitrisli,

Thank you for opening this PR as discussed in JIRA!
I tested your changes and they look good to me 👍 
I just requested I couple of minor changes (See comments above). Please don't hesitate also to update the date to 2018 in license headers of the files you changed.

Kind regards,
Mahmoud

**Comment by benas** [comment]
Awesome! Thank you for these updates.

Rebased, squashed and merged as f8f8a02a4d0ea111b638f9d48f3b3fae84fee200.

Thank you @dimitrisli for your contribution 👍

**Comment by dimitrisli** [comment]
@benas fantastic, many thanks!

I notice now that after the merge to master there is a test failing triggered by the Bamboo Build: https://build.spring.io/browse/BATCH-GRAD-JOB1-459 

I cannot reproduce locally (it always succeeds) so I'm a bit confused - let me know how can I help there.

Also if we wanted to back-port this change in 3.0.x I'm guessing we need another PR  to 3.0.x this time - shall I proceed towards that direction?

**Comment by benas** [comment]
The build failure was not related to your changes. The build on master is [successful](https://build.spring.io/browse/BATCH-GRAD/latest) now.

In regards to backports, I backported the fix in [4.0.x](https://github.com/spring-projects/spring-batch/commit/13ca0e0089e88bb460754192c4295080a4c4cb54) yesterday (build [successful](https://build.spring.io/browse/BATCH-GRAD2/latest)) and was about to do it for 3.0.x today. But if you want to help, you are welcome! So please go ahead and open a PR for 3.0.x, we will consider merging it when we release 3.0.10. Many thanks upfront!

**Pull request closed without merging**